File: Modul/SinglyLinkedList.py
from ListNode import ListNode
from ResepMakanan import ResepMakanan
class SinglyLinkedList():
    def __init__(self, listNama):
        self.__nama = listNama
        self.__firstNode = self.__lastNode = None
        
    def isEmpty(self):
        return bool(self.__firstNode == None)
        
    def InsertAtFront(self,insertItem):
        newNode = ListNode(insertItem)
        
        if self.isEmpty():
            self.__firstNode = self.__lastNode = newNode
        else:
            newNode.next = self.__firstNode
            self.__firstNode = newNode
            
    def InsertAtBack(self,insertItem):
        newNode = ListNode(insertItem)
        
        if self.isEmpty():
            self.__firstNode = self.__lastNode = newNode
        else:
            self.__lastNode.next = newNode
            self.__lastNode = newNode
            
    def removeFromFront(self):
        if self.isEmpty():
            raise Exception(f"{self.__nama} is empty")
        
        removedItem = self.__firstNode.data
        if (self.__firstNode == self.__lastNode):
            self.__firstNode = self.__lastNode = None
        else:
            self.__firstNode = self.__firstNode.next
        return removedItem
    
    def removeFromBack(self):
        if self.isEmpty():
            raise Exception(f"{self.__nama} is empty")
        
        removedItem = self.__lastNode.data
        if (self.__firstNode == self.__lastNode):
            self.__firstNode = self.__lastNode = None
        else:
            iterator = self.__firstNode
            while iterator.next != self.__lastNode:
                iterator = iterator.next
            self.__lastNode = iterator
            iterator.next = None
        return removedItem
    
    def print(self):
        if self.isEmpty():
            print(f"Empty {self.__nama}")
            return
        
        print(f"List {self.__nama} : ")
        
        iterator = self.__firstNode
        
        while iterator:
            print(str(iterator.data))
            iterator = iterator.next
            
    def cariResep(self, namaResep):
        iterator = self.__firstNode
        while iterator != None:
            if isinstance(iterator.data, ResepMakanan):
                if iterator.data.getNamaResep() == namaResep:
                    print(f"Resep ditemukan : {iterator.data}")
                    return
            iterator = iterator.next

File: Belajar/SinglyLinkedList.java
import java.util.NoSuchElementException;

public class SinglyLinkedList<E> {
    private ListNode<E> firstNode;
    private ListNode<E> lastNode;
    private String name;

    public SinglyLinkedList(String listName) {
        name = listName;
        firstNode = lastNode = null;
    }

    public void insertAtFront(E insertItem) {
        ListNode newNode = new ListNode(insertItem);
        if (isEmpty()) {
            firstNode = lastNode = new ListNode<E>(insertItem);
        } else {
            newNode.next = firstNode;
            firstNode = newNode;
        }
    }

    public void insertAtBack(E insertItem) {
        ListNode newNode = new ListNode(insertItem);

        if (isEmpty()) {
            firstNode = lastNode = new ListNode<E>(insertItem);
        } else {
            lastNode.next = newNode;
            lastNode = newNode;
        }
    }


    public E removeFromFront() throws NoSuchElementException {
        if (isEmpty()) {
            throw new NoSuchElementException(name + " is empty");
        }
        E removedItem = firstNode.data;
        if (firstNode == lastNode) {
            firstNode = lastNode = null;
        } else {
            firstNode = firstNode.next;
        }
        return removedItem;
    }

    public E removeFromBack() throws NoSuchElementException {
        if (isEmpty()) {
            throw new NoSuchElementException(name + " is empty");
        }
        E removedItem = lastNode.data;

        if (firstNode == lastNode) {
            firstNode = lastNode = null;
        } else {
            ListNode<E> iterator = firstNode;

            while (iterator.next != lastNode) {
                iterator = iterator.next;
            }
            lastNode = iterator;
            iterator.next = null;
        }

        return removedItem;
    }

    public boolean isEmpty() {
        return firstNode == null;
    }

    public void print() {
        if (isEmpty()) {
            System.out.printf("Empty %s%n", name);
            return;
        }

        System.out.printf("List %s: %n", name);

        ListNode<E> iterator = firstNode;

        while (iterator != null) {
            System.out.printf("%s", iterator.data);
            iterator = iterator.next;

            System.out.println();
        }
    }

    public void cariResep(String namaResep) {
        ListNode<E> current = firstNode;
        while (current != null) {
            if (current.data instanceof ResepMakanan) {
                ResepMakanan resep = (ResepMakanan) current.data;
                if (resep.getNamaResep().equals(namaResep)) {
                    System.out.println("Resep ditemukan: " + resep);
                    return;
                }
            }
            current = current.next;
        }
        System.out.println("Resep tidak ditemukan");
    }
}